File: code_samples/section6/problems/problem6_1/problem6_1.cs
// ===== Tests =====
//
// This file contains BOTH:
//   1) A lightweight "test harness" at the top (top-level statements),
//   2) A MinHeap implementation below.
//
// The tests are intentionally simple and print-oriented:
//   - Insert several values while showing the running minimum (Peek).
//   - Pop until empty to verify ascending (sorted) order.
//   - Confirm behavior when peeking/popping an empty heap (null results).
//   - Reinsert after emptying to confirm the heap remains usable.
//
// NOTE: This code uses modern C# features:
//   - Top-level statements (no explicit Main method).
//   - Collection expression syntax: int[] values = [ ... ];
//   - Index-from-end operator: data[^1]

static void Print(string label, IEnumerable<int> seq) =>
    // Utility method for printing a label followed by a sequence of ints.
    // Example output: "Values: 1 2 3"
    // This isn't used in the current test script, but can be handy
    // for quickly dumping arrays/lists during debugging.
    Console.WriteLine(label + string.Join(" ", seq));

Console.WriteLine("=== Testing MinHeap ===");

var h = new MinHeap();

// Sample values chosen to exercise multiple bubble-up operations and ordering.
// The resulting sorted pop sequence should be: 1 2 3 5 7 8
int[] values = [5, 3, 8, 1, 7, 2];

Console.Write("Inserting values: ");
foreach (var v in values) {
    // Print the value as we insert it
    Console.Write(v + " ");

    // Insert into heap (O(log n))
    h.Insert(v);

    // Peek returns the current minimum (O(1)); should track the smallest inserted so far
    Console.WriteLine($"\n  After insert {v} -> peek (min): {h.Peek()}");
}
Console.WriteLine();

// Pop repeatedly until the heap is empty.
// This should print values in ascending order if heap invariants are correct.
Console.Write("Extracting in sorted order: ");
while (true) {
    // Pop returns null when empty (instead of throwing).
    var x = h.Pop();
    if (x == null) break; // stop when heap is empty
    Console.Write(x + " ");
}
Console.WriteLine("\n");

// These calls demonstrate the "empty heap" behavior.
// Since Peek/Pop return nullable int (int?), we expect null outputs here.
Console.WriteLine("Testing peek/pop on empty heap:");
Console.WriteLine("peek(): " + h.Peek());
Console.WriteLine("pop(): " + h.Pop());
Console.WriteLine();

// Confirm the heap is reusable after becoming empty.
Console.WriteLine("Reinserting values: 10, 4, 6");
h.Insert(10);
h.Insert(4);
h.Insert(6);

// After inserting 10, 4, 6 the minimum should be 4.
Console.WriteLine($"peek(): {h.Peek()}");

// Pop again to verify ordering in a second run: expected 4 6 10.
Console.Write("Final pop sequence: ");
while (true) {
    var x = h.Pop();
    if (x == null) break;
    Console.Write(x + " ");
}
Console.WriteLine();

class MinHeap {
    /*
        MinHeap (priority queue) for integers backed by List<int>.

        Invariant:
            For every node i (except root), data[parent(i)] <= data[i]

        Array index mapping (0-based):
            parent(i) = (i - 1) / 2
            left(i)   = 2*i + 1
            right(i)  = 2*i + 2

        Complexity:
            Insert: O(log n)
            Peek:   O(1)
            Pop:    O(log n)

        Empty-heap behavior:
            - Peek() returns null when empty
            - Pop() returns null when empty

        Returning null avoids "sentinel value" ambiguity (e.g., INT_MIN),
        but it does mean callers must handle nullable results.
    */

    // Internal storage for heap elements in heap-order.
    // data[0] is always the minimum element (if non-empty).
    private readonly List<int> data = new List<int>();

    /*
        Swap
        ----
        Swaps elements at indices i and j.

        Uses tuple deconstruction assignment for a concise swap:
            (a, b) = (b, a);

        Assumes i and j are valid indices.
    */
    private void Swap(int i, int j) {
        (data[j], data[i]) = (data[i], data[j]);
    }

    /*
        BubbleUp
        --------
        Restores heap invariant after inserting an element.

        Insert places the new element at the end of the list.
        If the new element is smaller than its parent, we swap upward until:
            - we reach the root, or
            - parent <= child (heap property satisfied)

        i is the index of the newly inserted element.
    */
    private void BubbleUp(int i) {
        while (i > 0) {
            // Parent index in a binary heap stored in an array
            int p = (i - 1) / 2;

            // If current node is >= parent, heap property is satisfied
            if (data[i] >= data[p]) break;

            // Otherwise swap with parent and continue upward
            Swap(i, p);
            i = p;
        }
    }

    /*
        HeapifyDown
        -----------
        Restores heap invariant after removing the root (minimum).

        Pop typically:
            - replaces the root with the last element
            - removes the last element
            - then "pushes down" the moved root until heap property holds

        At each step:
            - compare current node with its children
            - swap with the smallest child if a child is smaller
            - stop when the current node is <= both children or has no children
    */
    private void HeapifyDown(int i) {
        int n = data.Count; // snapshot of size for bounds checks

        while (true) {
            int smallest = i;
            int left = 2 * i + 1;
            int right = 2 * i + 2;

            // Select the smallest among (i, left child, right child), if children exist
            if (left < n && data[left] < data[smallest]) smallest = left;
            if (right < n && data[right] < data[smallest]) smallest = right;

            // If i is already the smallest, the heap property holds here
            if (smallest == i) break;

            // Swap down and continue from the child we swapped with
            Swap(i, smallest);
            i = smallest;
        }
    }

    /*
        Insert
        ------
        Adds a value to the heap.

        Steps:
            1) Append to the end of the list
            2) BubbleUp from that index to restore heap property
    */
    public void Insert(int x) {
        data.Add(x);
        BubbleUp(data.Count - 1);
    }

    /*
        Peek
        ----
        Returns the current minimum (root) WITHOUT removing it.

        Returns:
            - null if heap is empty
            - otherwise data[0]

        This is O(1).
    */
    public int? Peek() {
        return data.Count == 0 ? (int?)null : data[0];
    }

    /*
        Pop
        ---
        Removes and returns the current minimum (root).

        Cases:
            1) Empty heap:
                - return null
            2) Single-element heap:
                - remove and return that element
            3) General case:
                - save root
                - move last element to root
                - remove last element
                - heapifyDown from root
                - return saved root

        Complexity: O(log n) in the general case.
    */
    public int? Pop() {
        if (data.Count == 0) return null;

        // Fast path: one element only
        if (data.Count == 1) {
            int v = data[0];

            // RemoveAt(0) is fine here because count==1 (no shifting cost concerns)
            data.RemoveAt(0);
            return v;
        }

        // Save the minimum value to return
        int root = data[0];

        // Move the last element into the root position
        data[0] = data[^1]; // ^1 = last element

        // Remove the last element (now duplicated at the root)
        data.RemoveAt(data.Count - 1);

        // Restore heap property starting from root
        HeapifyDown(0);

        return root;
    }
}

File: code_samples/section6/problems/problem6_1/problem6_1.py
"""
Problem 6.1 - Implement a Min-Heap (Array-Based)

This module contains:
  1) A simple array-backed MinHeap implementation for integers
  2) A small print-based test harness at the bottom

A min-heap maintains the invariant:
    For every node i (except the root),
        heap[parent(i)] <= heap[i]

Because of this:
  - The minimum element is always at index 0
  - insert is O(log n) via "bubble up"
  - pop (remove-min) is O(log n) via "heapify down"
  - peek is O(1)

Array index mapping (0-based):
  parent(i) = (i - 1) // 2
  left(i)   = 2*i + 1
  right(i)  = 2*i + 2

Empty-heap behavior:
  - peek() returns None when the heap is empty
  - pop() returns None when the heap is empty
"""

from typing import Optional


class MinHeap:
    """
    Min heap implementation (integers) backed by a Python list.

    Internal representation:
      - self.data is a list that stores elements in "heap order"
      - self.data[0] is always the minimum element (when non-empty)

    Note:
      - This implementation relies on integer comparisons.
      - For custom ordering or objects, you'd typically accept a key/comparator.
    """

    def __init__(self) -> None:
        # Backing storage for the heap. Starts empty and grows as needed.
        self.data: list[int] = []

    def _bubble_up(self, i: int) -> None:
        """
        Restore heap invariant after inserting an element.

        When we insert, we append the new element at the end of the list.
        That new element may be smaller than its parent, violating the heap
        invariant. "Bubble up" repeatedly swaps the element with its parent
        until the invariant is restored.

        Args:
            i: The index of the element to bubble upward (typically the last index).

        Complexity:
            O(log n) in the worst case (height of the heap).
        """
        while i > 0:
            # Compute parent index in a 0-based binary heap array
            p = (i - 1) // 2

            # If parent <= current, heap property is satisfied; stop.
            if self.data[i] >= self.data[p]:
                break

            # Otherwise swap current with parent and continue upward.
            self.data[i], self.data[p] = self.data[p], self.data[i]
            i = p

    def _heapify(self, i: int) -> None:
        """
        Restore heap invariant after removing the root (min element).

        Typical pop strategy:
          1) Save the root (minimum).
          2) Move the last element into index 0.
          3) Remove the last element.
          4) The moved element may be larger than its children, so we "heapify down"
             (a.k.a. sift down) by swapping it with the smaller child until the
             heap invariant is restored.

        Args:
            i: Index to heapify down from (usually 0).

        Complexity:
            O(log n) in the worst case.
        """
        n = len(self.data)

        while True:
            smallest = i
            left = 2 * i + 1
            right = 2 * i + 2

            # Pick the smallest among current node and its children (if they exist).
            if left < n and self.data[left] < self.data[smallest]:
                smallest = left
            if right < n and self.data[right] < self.data[smallest]:
                smallest = right

            # If the current node is already the smallest, heap property holds here.
            if smallest == i:
                break

            # Otherwise swap down and continue from where we swapped to.
            self.data[i], self.data[smallest] = self.data[smallest], self.data[i]
            i = smallest

    def insert(self, val: int) -> None:
        """
        Insert an element into the heap.

        Steps:
          1) Append the value to the end of the array
          2) Bubble it up to restore the heap property

        Args:
            val: Value to insert.

        Complexity:
            O(log n).
        """
        self.data.append(val)
        self._bubble_up(len(self.data) - 1)

    def peek(self) -> Optional[int]:
        """
        Return the minimum element without removing it.

        Returns:
            The minimum element (root) if present; otherwise None.

        Complexity:
            O(1).
        """
        return self.data[0] if self.data else None

    def pop(self) -> Optional[int]:
        """
        Remove and return the minimum element.

        Cases:
          - Empty heap: return None
          - One element: pop and return it
          - Otherwise:
              * save root
              * move last element to root
              * heapify down
              * return saved root

        Returns:
            The minimum element, or None if the heap is empty.

        Complexity:
            O(log n) when non-empty.
        """
        if not self.data:
            return None

        # Fast path: only one element
        if len(self.data) == 1:
            return self.data.pop()

        # Save min element
        root = self.data[0]

        # Move last element to root and shrink list
        self.data[0] = self.data.pop()

        # Restore heap property
        self._heapify(0)

        return root


# ===== Tests =====
#
# Simple print-driven checks:
#   - Insert values and show current minimum after each insert
#   - Pop until empty to verify ascending order
#   - Confirm peek/pop return None on empty
#   - Reinsert after emptying and pop again

def print_array(label: str, arr: list[int]) -> None:
    """
    Utility function to print the contents of a list.

    Note: Included for convenience/debugging. Not required by the heap itself.
    """
    print(label + " ".join(str(x) for x in arr))


print("=== Testing MinHeap ===")

h = MinHeap()
values = [5, 3, 8, 1, 7, 2]

print("Inserting values:", end=" ")
for v in values:
    print(v, end=" ")
    h.insert(v)

    # After each insert, peek should return the smallest value inserted so far
    print(f"\n  After insert {v} -> peek (min): {h.peek()}")
print()

# Repeated pop should yield values in ascending order if heap is correct.
print("Extracting in sorted order:", end=" ")
while True:
    x = h.pop()
    if x is None:
        break
    print(x, end=" ")
print("\n")

# Confirm empty-heap behavior
print("Testing peek/pop on empty heap:")
print("peek():", h.peek())
print("pop():", h.pop())
print()

# Confirm heap can be reused
print("Reinserting values: 10, 4, 6")
h.insert(10)
h.insert(4)
h.insert(6)

print("peek():", h.peek())

print("Final pop sequence:", end=" ")
while True:
    x = h.pop()
    if x is None:
        break
    print(x, end=" ")
print()
